runtime/src/main/java/org/corfudb/runtime/collections/streaming/StreamPollingScheduler.java

```diff
@@ -0,0 +1,321 @@
+            Preconditions.checkState(res.size() == batch.size());
+            allQueryResults.putAll(res);
+        }
+
+        Preconditions.checkState(tasks.size() == queries.size());
+
+        for (int idx = 0; idx < tasks.size(); idx++) {
+            StreamingTask task = tasks.get(idx);
+            StreamAddressRange taskQuery = queries.get(idx);
+            Preconditions.checkState(task.getStream().getStreamId().equals(taskQuery.getStreamID()));
+            Preconditions.checkState(allQueryResults.containsKey(taskQuery.getStreamID()),
+                    "StreamAddressSpace missing for %s", task.getStream().getStreamId());
+
+            StreamAddressSpace sas = allQueryResults.get(task.getStream().getStreamId()).getAddressesInRange(taskQuery);
+            task.getStream().refresh(sas);
+        }
+    }
+
+    @Data
+    class Interval {
+        private final long x;
+        private final long y;
+    }
+
+    private Interval merge(Interval interval, StreamAddressRange range) {
+        long newMinX = Math.min(interval.x, range.getEnd());
+        long newMaxY = Math.max(interval.y, range.getStart());
+        return new Interval(newMinX, newMaxY);
+    }
+
+    // Generate a range per stream that contains all sub-ranges that need to be queries
+
+    /**
+     * It's possible that multiple streaming tasks need to query different/overlapping ranges
+     * of the same stream. In order to prevent issuing multiple queries for the same stream,
+     * all range queries for the same stream are coalesced into a single request.
+     * For example, the following queries will all be merged into one query (i.e., (-1, 100])
+     *  Query1  (-1, 100)
+     *  Query2  (2, 5)
+     *  Query3  (3, 10)
+     */
+    private List<StreamAddressRange> coalesce(List<StreamAddressRange> ranges) {
+        Map<UUID, Interval> adjusted = new HashMap<>();
+
+        for (StreamAddressRange range : ranges) {
+            Interval interval = adjusted.getOrDefault(range.getStreamID(), new Interval(Long.MAX_VALUE, Long.MIN_VALUE));
+            Interval expandedInterval = merge(interval, range);
+            Preconditions.checkState(expandedInterval.x >= Address.NEVER_READ);
+            Preconditions.checkState(expandedInterval.x < expandedInterval.y);
+
+            adjusted.put(range.getStreamID(), expandedInterval);
+        }
+
+        List<StreamAddressRange> adjustedList = new ArrayList<>();
+        for (Map.Entry<UUID, Interval> entry : adjusted.entrySet()) {
+            adjustedList.add(new StreamAddressRange(entry.getKey(), entry.getValue().y, entry.getValue().x));
+        }
+        return adjustedList;
+    }
+
+    private void dispatchSyncTasks(List<StreamingTask> tasks) {
+        for (StreamingTask task : tasks) {
+            if (task.getStream().hasNext()) {
+                // this will fail for tasks that are already syncing
```

Comment:
doesn't that mean that it will break out for all other tasks which could still be submitted to the workers executor?

Reply:
Its by design. I want it to fail the scheduler thread explicitly to make clear any concurrency issues. In other words, this is a fault that cannot be ignored.